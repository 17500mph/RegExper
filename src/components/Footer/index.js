import React from 'react';
import PropTypes from 'prop-types';
import { StaticQuery, graphql } from 'gatsby';
import { withNamespaces, Trans } from 'react-i18next';

import style from './style.module.css';

const query = graphql`
  query FooterQuery {
    site {
      siteMetadata {
        buildId
      }
    }
  }
`;

export const Footer = ({ t, buildId }) => (
  <footer className={ style.footer }>
    <ul className={ style.list }>
      <li>
        <Trans>Created by <a
          href="mailto:jeff.avallone@gmail.com">Jeff Avallone</a></Trans>
      </li>
      <li>
        <Trans>Generated images licensed: <a
          href="http://creativecommons.org/licenses/by/3.0/"
          rel="license external noopener noreferrer"
          target="_blank">
          <img src="https://licensebuttons.net/l/by/3.0/80x15.png"
            alt={ t('Creative Commons CC-BY-3.0 License') } />
        </a></Trans>
      </li>
    </ul>
    <div className={ style.buildId }>
      { buildId }
    </div>
  </footer>
);

Footer.propTypes = {
  t: PropTypes.func.isRequired,
  buildId: PropTypes.string.isRequired
};

export default withNamespaces()(props => (
  <StaticQuery query={ query } render={ ({ site: { siteMetadata } }) => (
    <Footer { ...props } { ...siteMetadata } />
  ) } />
));
