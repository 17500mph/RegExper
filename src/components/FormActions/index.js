import React from 'react';
import PropTypes from 'prop-types';
import { withNamespaces, Trans } from 'react-i18next';

import DownloadIcon from 'react-feather/dist/icons/download';
import LinkIcon from 'react-feather/dist/icons/link';

import style from './style.module.css';

import { createPngLink, createSvgLink } from './links';

class FormActions extends React.PureComponent {
  static propTypes = {
    permalinkUrl: PropTypes.string,
    imageDetails: PropTypes.shape({
      svg: PropTypes.string,
      width: PropTypes.number,
      height: PropTypes.number
    }),
    t: PropTypes.func.isRequired
  }

  state = {
    svgLink: null,
    pngLink: null
  }

  componentDidMount() {
    const { imageDetails } = this.props;

    if (imageDetails && imageDetails.svg) {
      this.generateDownloadLinks();
    }
  }

  componentDidUpdate(prevProps) {
    const { imageDetails } = this.props;
    const { imageDetails: prevImageDetails } = prevProps;

    if (!imageDetails) {
      this.setState({ svgLink: null, pngLink: null });
      return;
    }

    if (!prevImageDetails) {
      this.generateDownloadLinks();
      return;
    }

    if (imageDetails.svg !== prevImageDetails.svg
      || imageDetails.width !== prevImageDetails.width
      || imageDetails.height !== prevImageDetails.height) {
      this.generateDownloadLinks();
      return;
    }
  }

  async generateDownloadLinks() {
    const { imageDetails: { svg, width, height } } = this.props;

    this.setState({
      svgLink: await createSvgLink({ svg }),
      pngLink: await createPngLink({ svg, width, height })
    });
  }

  downloadLink({ url, filename, type, label }) {
    const { t } = this.props;

    return <li>
      <a href={ url } download={ filename } type={ type }>
        <DownloadIcon />{ t(label) }
      </a>
    </li>;
  }

  render() {
    const {
      permalinkUrl
    } = this.props;
    const {
      svgLink,
      pngLink
    } = this.state;

    return <ul className={ style.actions }>
      { pngLink && this.downloadLink(pngLink) }
      { svgLink && this.downloadLink(svgLink) }
      { permalinkUrl && <li>
        <a href={ permalinkUrl }><LinkIcon /><Trans>Permalink</Trans></a>
      </li> }
    </ul>;
  }
}

export { FormActions };
export default withNamespaces()(FormActions);
