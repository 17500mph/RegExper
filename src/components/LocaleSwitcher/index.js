import React from 'react';
import { withTranslation, Trans } from 'react-i18next';

import ExpandIcon from 'react-feather/dist/icons/chevrons-down';

import i18n, { locales } from 'i18n';

import localeToAvailable from './locale-to-available';
import style from './style.module.css';

export class LocaleSwitcher extends React.PureComponent {
  state = {
    current: localeToAvailable(
      i18n.language || '',
      locales.map(l => l.code),
      'en')
  }

  componentDidMount() {
    i18n.on('languageChanged', this.handleLanguageChange);
  }

  componentWillUnmount() {
    i18n.off('languageChanged', this.handleLanguageChange);
  }

  handleSelectChange = ({ target }) => {
    i18n.changeLanguage(target.value);
  }

  handleLanguageChange = lang => {
    this.setState({ current: lang });
  }

  render() {
    const { current } = this.state;

    return <label>
      <Trans>Language</Trans>
      <div className={ style.switcher }>
        <select value={ current } onChange={ this.handleSelectChange }>
          { locales.map(locale => (
            <option value={ locale.code } key={ locale.code }>
              { locale.name }
            </option>
          )) }
        </select>
        <ExpandIcon />
      </div>
    </label>;
  }
}

export default withTranslation()(LocaleSwitcher);
