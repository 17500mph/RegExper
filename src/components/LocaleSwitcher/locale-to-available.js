const localeToAvailable = (locale, available, defaultLocale) => {
  if (available.includes(locale)) {
    return locale;
  }

  const parts = locale.split('-');

  if (parts.length > 0 && available.includes(parts[0])) {
    return parts[0];
  }

  return defaultLocale;
};

export default localeToAvailable;
