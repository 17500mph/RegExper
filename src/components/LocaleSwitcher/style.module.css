@import url('../../globals.module.css');

:root {
  --control-gradient: var(--color-tan) var(--gradient-tan);
  --select-height: 2.4rem;
  --select-width: 10rem;
}

.switcher {
  composes: fancy-select;
}
