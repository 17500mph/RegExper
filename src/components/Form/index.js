import React from 'react';

import ExpandIcon from 'react-feather/dist/icons/chevrons-down';

import style from './style.module.css';

import AppContext from 'components/App/context';
import FormActions from 'components/FormActions';

const syntaxList = [
  { id: 'js', label: 'JavaScript' },
  { id: 'pcre', label: 'PCRE' }
];

class Form extends React.PureComponent {
  static contextType = AppContext

  state = {
    expr: this.context.expr,
    syntax: this.context.syntax
  }

  componentDidUpdate() {
    if (this.state.expr === undefined && this.state.syntax === undefined) {
      this.setState({
        syntax: this.context.syntax,
        expr: this.context.expr
      });
    }
  }

  handleSubmit = event => {
    event.preventDefault();

    const { expr, syntax } = this.state;

    this.context.renderExpr({ expr, syntax });
  }

  handleKeyPress = event => {
    if (event.charCode === 13 && event.shiftKey) {
      this.handleSubmit(event);
    }
  }

  handleChange = event => this.setState({
    [event.target.name]: event.target.value
  })

  render() {
    const { expr, syntax } = this.state;

    return <div className={ style.form }>
      <form onSubmit={ this.handleSubmit }>
        <textarea
          name="expr"
          value={ expr }
          onKeyPress={ this.handleKeyPress }
          onChange={ this.handleChange }
          autoFocus
          placeholder="Enter regular expression to display"></textarea>
        <button type="submit">Display</button>
        <div className={ style.select }>
          <select
            name="syntax"
            value={ syntax }
            onChange={ this.handleChange } >
            { syntaxList.map(({ id, label }) => (
              <option value={ id } key={ id }>{ label }</option>
            )) }
          </select>
          <ExpandIcon />
        </div>
        <FormActions />
      </form>
    </div>;
  }
}

export default Form;
