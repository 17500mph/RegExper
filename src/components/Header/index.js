import React from 'react';
import { translate, Trans } from 'react-i18next';

import style from './style.css';
import GitlabIcon from 'feather-icons/dist/icons/gitlab.svg';

import LocaleSwitcher from 'components/LocaleSwitcher';

const Header = () => (
  <header className={ style.header } data-banner={ process.env.BANNER }>
    <h1>
      <a href="/">Regexper</a>
    </h1>

    <ul>
      <li><a href="https://gitlab.com/javallone/regexper-static" rel="external noopener noreferrer" target="_blank">
        <GitlabIcon/><Trans>Source on GitLab</Trans>
      </a></li>
      <li>
        <a href="/privacy.html"><Trans>Privacy Policy</Trans></a>
      </li>
      <li><LocaleSwitcher /></li>
    </ul>
  </header>
);

export default translate()(Header);
export { Header };
