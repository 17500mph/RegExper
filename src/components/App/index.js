import React from 'react';
import PropTypes from 'prop-types';
import * as Sentry from '@sentry/browser';
import URLSearchParams from '@ungap/url-search-params';

import Form from 'components/Form';
import FormActions from 'components/FormActions';
import Loader from 'components/Loader';
import Message from 'components/Message';

class App extends React.PureComponent {
  state = {
    loading: false,
    loadingError: null,
    Render: null
  }

  componentDidMount() {
    if (this.props.expr) {
      this.handleRender();
    }
  }

  componentDidUpdate(prevProps) {
    const { syntax, expr } = this.props;

    if (syntax !== prevProps.syntax || expr !== prevProps.expr) {
      this.handleRender();
    }
  }

  handleSubmit = ({ syntax, expr }) => {
    if (expr) {
      document.location.hash = new URLSearchParams({
        syntax,
        expr
      }).toString();
    }
  }

  handleRender = async () => {
    const { syntax, expr } = this.props;

    this.setState({
      loading: false,
      loadingError: null,
      Render: null
    });

    if (!expr) {
      return;
    }

    this.setState({
      loading: true
    });

    try {
      const Render = await import(
        /* webpackChunkName: "render-[index]" */
        `syntax/${ syntax }`
      );

      // HACK: Fake loading time
      await new Promise(resolve => setTimeout(resolve, 2000));

      this.setState({
        loading: false,
        Render: Render.default
      });
    }
    catch (e) {
      Sentry.withScope(scope => {
        scope.setExtra('syntax', syntax);
        Sentry.captureException(e);
      });
      this.setState({
        loading: false,
        loadingError: e
      });
      throw e;
    }
  }

  handleRetry = event => {
    event.preventDefault();
    this.handleRender();
  }

  handleSvg = imageDetails => this.setState({ imageDetails });

  render() {
    const {
      syntax,
      expr,
      permalinkUrl,
      syntaxList
    } = this.props;
    const {
      loading,
      loadingError,
      Render,
      imageDetails
    } = this.state;

    const formProps = {
      onSubmit: this.handleSubmit,
      syntax,
      expr,
      syntaxList
    };
    const actionProps = {
      imageDetails,
      permalinkUrl
    };
    const renderProps = {
      onRender: this.handleSvg,
      expr
    };

    return <>
      <Form { ...formProps }>
        { Render && <FormActions { ...actionProps } /> }
      </Form>

      { loading && <Loader /> }

      { loadingError && <Message type="error" heading="Render Failure">
        <p>An error occurred while rendering the regular expression.</p>
        <a href="#retry" onClick={ this.handleRetry }>Retry</a>
      </Message> }

      { Render && <Render { ...renderProps } /> }
    </>;
  }
}

App.propTypes = {
  syntax: PropTypes.string,
  expr: PropTypes.string,
  permalinkUrl: PropTypes.string,
  syntaxList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    label: PropTypes.string
  }))
};

export default App;
