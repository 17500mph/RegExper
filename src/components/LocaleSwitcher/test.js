import React from 'react';
import { shallow } from 'enzyme';

import i18n from 'i18n';
import { LocaleSwitcherImpl } from 'components/LocaleSwitcher';

describe('LocaleSwitcher', () => {
  test('rendering', () => {
    const component = shallow(
      <LocaleSwitcherImpl />
    );
    expect(component).toMatchSnapshot();
  });

  test('changing language', () => {
    jest.spyOn(i18n, 'changeLanguage');

    const component = shallow(
      <LocaleSwitcherImpl />
    );
    const selectInput = component.find('select');
    selectInput.value = 'other';
    selectInput.simulate('change', { target: { value: 'other' } });

    expect(i18n.changeLanguage).toHaveBeenCalledWith('other');
  });
});
