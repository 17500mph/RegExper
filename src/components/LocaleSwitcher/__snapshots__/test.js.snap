// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleSwitcher rendering 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LocaleSwitcherImpl />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <WithMergedOptions(TransComponent)>
          Language
        </WithMergedOptions(TransComponent)>,
        <div
          className="switcher"
        >
          <select
            onChange={[Function]}
            value="en"
          >
            <option
              value="en"
            >
              English
            </option>
            <option
              value="other"
            >
              Other
            </option>
          </select>
          <ChevronsDown
            color="currentColor"
            size="24"
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Language",
        },
        "ref": null,
        "rendered": "Language",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <select
              onChange={[Function]}
              value="en"
            >
              <option
                value="en"
              >
                English
              </option>
              <option
                value="other"
              >
                Other
              </option>
            </select>,
            <ChevronsDown
              color="currentColor"
              size="24"
            />,
          ],
          "className": "switcher",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  value="en"
                >
                  English
                </option>,
                <option
                  value="other"
                >
                  Other
                </option>,
              ],
              "onChange": [Function],
              "value": "en",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "en",
                "nodeType": "host",
                "props": Object {
                  "children": "English",
                  "value": "en",
                },
                "ref": null,
                "rendered": "English",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": "other",
                "nodeType": "host",
                "props": Object {
                  "children": "Other",
                  "value": "other",
                },
                "ref": null,
                "rendered": "Other",
                "type": "option",
              },
            ],
            "type": "select",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "color": "currentColor",
              "size": "24",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "label",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <WithMergedOptions(TransComponent)>
            Language
          </WithMergedOptions(TransComponent)>,
          <div
            className="switcher"
          >
            <select
              onChange={[Function]}
              value="en"
            >
              <option
                value="en"
              >
                English
              </option>
              <option
                value="other"
              >
                Other
              </option>
            </select>
            <ChevronsDown
              color="currentColor"
              size="24"
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Language",
          },
          "ref": null,
          "rendered": "Language",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <select
                onChange={[Function]}
                value="en"
              >
                <option
                  value="en"
                >
                  English
                </option>
                <option
                  value="other"
                >
                  Other
                </option>
              </select>,
              <ChevronsDown
                color="currentColor"
                size="24"
              />,
            ],
            "className": "switcher",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option
                    value="en"
                  >
                    English
                  </option>,
                  <option
                    value="other"
                  >
                    Other
                  </option>,
                ],
                "onChange": [Function],
                "value": "en",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "en",
                  "nodeType": "host",
                  "props": Object {
                    "children": "English",
                    "value": "en",
                  },
                  "ref": null,
                  "rendered": "English",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": "other",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Other",
                    "value": "other",
                  },
                  "ref": null,
                  "rendered": "Other",
                  "type": "option",
                },
              ],
              "type": "select",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "color": "currentColor",
                "size": "24",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "label",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
